mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/BlockMatrix.scala

```diff
@@ -0,0 +1,242 @@
+    nCols
+  }
+
+  /** Returns the dimensions of the matrix. */
+  private def getDim: (Long, Long) = {
+    case class MatrixMetaData(var rowIndex: Int, var colIndex: Int,
+        var numRows: Int, var numCols: Int)
+    // picks the sizes of the matrix with the maximum indices
+    def pickSizeByGreaterIndex(example: MatrixMetaData, base: MatrixMetaData): MatrixMetaData = {
+      if (example.rowIndex > base.rowIndex) {
+        base.rowIndex = example.rowIndex
+        base.numRows = example.numRows
+      }
+      if (example.colIndex > base.colIndex) {
+        base.colIndex = example.colIndex
+        base.numCols = example.numCols
+      }
+      base
+    }
+
+    // Aggregate will return an error if the rdd is empty
+    val lastRowCol = rdd.treeAggregate(new MatrixMetaData(0, 0, 0, 0))(
+      seqOp = (c, v) => (c, v) match { case (base, ((blockXInd, blockYInd), mat)) =>
+        pickSizeByGreaterIndex(
+          new MatrixMetaData(blockXInd, blockYInd, mat.numRows, mat.numCols), base)
+      },
+      combOp = (c1, c2) => (c1, c2) match {
+        case (res1, res2) =>
+          pickSizeByGreaterIndex(res1, res2)
+      })
+    // We add the size of the edge matrices, because they can be less than the specified
+    // rowsPerBlock or colsPerBlock.
+    (lastRowCol.rowIndex.toLong * rowsPerBlock + lastRowCol.numRows,
+      lastRowCol.colIndex.toLong * colsPerBlock + lastRowCol.numCols)
+  }
+
+  /** Returns the Frobenius Norm of the matrix */
+  def normFro(): Double = {
+    math.sqrt(rdd.map { mat => mat._2 match {
+      case sparse: SparseMatrix =>
+        sparse.values.map(x => math.pow(x, 2)).sum
+      case dense: DenseMatrix =>
+        dense.values.map(x => math.pow(x, 2)).sum
+      }
+    }.reduce(_ + _))
+  }
+
+  /** Cache the underlying RDD. */
+  def cache(): BlockMatrix = {
+    rdd.cache()
+    this
+  }
+
+  /** Set the storage level for the underlying RDD. */
+  def persist(storageLevel: StorageLevel): BlockMatrix = {
+    rdd.persist(storageLevel)
+    this
+  }
+
+  /** Collect the distributed matrix on the driver as a `DenseMatrix`. */
+  def toLocalMatrix(): Matrix = {
+    require(numRows() < Int.MaxValue, "The number of rows of this matrix should be less than " +
+      s"Int.MaxValue. Currently numRows: ${numRows()}")
+    require(numCols() < Int.MaxValue, "The number of columns of this matrix should be less than " +
+      s"Int.MaxValue. Currently numCols: ${numCols()}")
+    val nRows = numRows().toInt
+    val nCols = numCols().toInt
+    val mem = nRows * nCols * 8 / 1000000
+    if (mem > 500) logWarning(s"Storing this matrix will require $mem MB of memory!")
+
+    val parts = rdd.collect().sortBy(x => (x._1._2, x._1._1))
+    val values = new Array[Double](nRows * nCols)
+    parts.foreach { case ((rowIndex, colIndex), block) =>
+      val rowOffset = rowIndex * rowsPerBlock
+      val colOffset = colIndex * colsPerBlock
+      var j = 0
+      val mat = block.toArray
```

Comment:
I can modify it in the next 3 upcoming PRs (matrix multiplication-addition, conversions, repartition) if it doesn't.